circleci: failed tests: TestRemoveRangeWithoutGC

The following test appears to have failed:

[#23021](https://circleci.com/gh/cockroachdb/cockroach/23021):

```
I160922 16:46:37.439959 14158 rpc/context.go:181  closing 127.0.0.1:46098
I160922 16:46:37.439975 14107 rpc/context.go:181  closing 127.0.0.1:60951
W160922 16:46:37.440029 14215 storage/raft_transport.go:433  raft transport stream to node 1 failed: rpc error: code = 13 desc = transport is closing
I160922 16:46:37.440185 14292 storage/engine/rocksdb.go:404  closing in-memory rocksdb instance
I160922 16:46:37.440460 14292 storage/engine/rocksdb.go:404  closing in-memory rocksdb instance
--- FAIL: TestRemoveRangeWithoutGC (45.14s)
    testing.go:117: storage/client_raft_test.go:2311, condition failed to evaluate within 45s: expected replica to be missing
=== RUN   TestCheckConsistencyMultiStore
I160922 16:46:37.441350 14295 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I160922 16:46:37.443420 14295 gossip/gossip.go:230  initial resolvers: []
W160922 16:46:37.443459 14295 gossip/gossip.go:1041  no resolvers found; use --join to specify a connected node
I160922 16:46:37.443613 14295 gossip/gossip.go:264  NodeID set to 1
I160922 16:46:37.449425 14295 rpc/context.go:215  dialing 127.0.0.1:38104
I160922 16:46:37.449917 14295 gossip/gossip.go:264  NodeID set to 1
I160922 16:46:37.450607 14295 storage/store.go:998  [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I160922 16:46:37.450706 14295 gossip/gossip.go:278  NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:38104" > attrs:<> 
--
I160922 16:46:52.712250 1 stopper.go:419  quiesceing; tasks left:
2      gossip/infostore.go:294
I160922 16:46:52.712291 1 stopper.go:419  quiesceing; tasks left:
1      gossip/infostore.go:294
--- PASS: Example_rebalancing (0.16s)
FAIL
FAIL    github.com/cockroachdb/cockroach/storage    75.145s
=== RUN   TestBatchBasics
I160922 16:46:14.075863 9 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I160922 16:46:14.078436 9 storage/engine/rocksdb.go:404  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN   TestBatchRepr
I160922 16:46:14.081504 19 storage/engine/rocksdb.go:356  opening in memory rocksdb instance
I160922 16:46:14.082254 19 storage/engine/rocksdb.go:404  closing in-memory rocksdb instance
--- PASS: TestBatchRepr (0.00s)
=== RUN   TestBatchGet
I160922 16:46:14.084241 21 storage/engine/rocksdb.go:356  opening in memory rocksdb instance

```

Please assign, take a look and update the issue accordingly.

Easily reproduced in GCE with

```
make stress STRESSFLAGS='-p 100 -stderr -maxfails 1' PKG=./storage TESTS=TestRemoveRangeWithoutGC
```

```
627 runs so far, 0 failures, over 5s
1270 runs so far, 0 failures, over 10s
1874 runs so far, 0 failures, over 15s
2459 runs so far, 0 failures, over 20s
3017 runs so far, 0 failures, over 25s
3567 runs so far, 0 failures, over 30s
4110 runs so far, 0 failures, over 35s
4633 runs so far, 0 failures, over 40s
5149 runs so far, 0 failures, over 45s
5664 runs so far, 0 failures, over 50s
6164 runs so far, 0 failures, over 55s
6670 runs so far, 0 failures, over 1m0s
7160 runs so far, 0 failures, over 1m5s
7656 runs so far, 0 failures, over 1m10s
8143 runs so far, 0 failures, over 1m15s
8627 runs so far, 0 failures, over 1m20s
9109 runs so far, 0 failures, over 1m25s
9596 runs so far, 0 failures, over 1m30s
10069 runs so far, 0 failures, over 1m35s
10544 runs so far, 0 failures, over 1m40s
11026 runs so far, 0 failures, over 1m45s
11501 runs so far, 0 failures, over 1m50s
11973 runs so far, 0 failures, over 1m55s
12440 runs so far, 0 failures, over 2m0s
12913 runs so far, 0 failures, over 2m5s
13385 runs so far, 0 failures, over 2m10s
13861 runs so far, 0 failures, over 2m15s
14331 runs so far, 0 failures, over 2m20s
14803 runs so far, 0 failures, over 2m25s
15277 runs so far, 0 failures, over 2m30s
15745 runs so far, 0 failures, over 2m35s
16209 runs so far, 0 failures, over 2m40s
16681 runs so far, 0 failures, over 2m45s
17140 runs so far, 0 failures, over 2m50s
17613 runs so far, 0 failures, over 2m55s
18075 runs so far, 0 failures, over 3m0s
18541 runs so far, 0 failures, over 3m5s
19010 runs so far, 0 failures, over 3m10s
19471 runs so far, 0 failures, over 3m15s

I161004 19:12:21.804495 1 rand.go:76  Random seed: 7408785237016292379
I161004 19:12:21.857920 11 storage/engine/rocksdb.go:360  opening in memory rocksdb instance
I161004 19:12:21.865449 11 gossip/gossip.go:234  initial resolvers: []
W161004 19:12:21.865547 11 gossip/gossip.go:1045  no resolvers found; use --join to specify a connected node
I161004 19:12:21.865769 11 gossip/gossip.go:268  NodeID set to 1
I161004 19:12:22.045882 11 gossip/gossip.go:268  NodeID set to 1
I161004 19:12:22.058723 11 storage/store.go:1040  [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I161004 19:12:22.058875 11 gossip/gossip.go:282  NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:40629" > attrs:<>
I161004 19:12:22.060925 146 storage/replica_trigger.go:252  [s1,r1:/M{in-ax}] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 900ms following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=1970-01-01 00:00:00 +0000 UTC]
I161004 19:12:22.061505 11 storage/engine/rocksdb.go:360  opening in memory rocksdb instance
I161004 19:12:22.064405 11 gossip/gossip.go:234  initial resolvers: [127.0.0.1:40629]
W161004 19:12:22.064432 11 gossip/gossip.go:1047  no incoming or outgoing connections
I161004 19:12:22.064475 11 gossip/gossip.go:268  NodeID set to 2
I161004 19:12:22.067451 47 gossip/client.go:125  node 2: started gossip client to 127.0.0.1:40629
I161004 19:12:22.073806 11 gossip/gossip.go:268  NodeID set to 2
I161004 19:12:22.077623 11 storage/store.go:1040  [n2,s2]: failed initial metrics computation: [n2,s2]: system config not yet available
I161004 19:12:22.080648 11 gossip/gossip.go:282  NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:40067" > attrs:<>
I161004 19:12:22.132597 11 storage/replica_raftstorage.go:443  [s1,r1:/M{in-ax}] generated snapshot 73e4c554 for range 1 at index 12 in 95.542µs.
I161004 19:12:22.133374 11 storage/store.go:2747  streamed snapshot: kv pairs: 31, log entries: 2
I161004 19:12:22.133840 72 storage/replica_raftstorage.go:573  [s2] [n2,s2,r1:{-}]: with replicaID [?], applying preemptive snapshot at index 12 (id=73e4c554, encoded size=16, 1 rocksdb batches, 2 log entries)
I161004 19:12:22.134633 72 storage/replica_raftstorage.go:576  [s2] [n2,s2,r1:/M{in-ax}]: with replicaID [?], applied preemptive snapshot in 0.001s
I161004 19:12:22.135458 11 storage/replica_command.go:3236  change replicas: read existing descriptor range_id:1 start_key:"" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2
I161004 19:12:22.136799 371 storage/replica.go:1712  [s1,r1:/M{in-ax}] proposing ADD_REPLICA {NodeID:2 StoreID:2 ReplicaID:2} for range 1: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2}]
I161004 19:12:22.138636 11 storage/replica_command.go:3236  change replicas: read existing descriptor range_id:1 start_key:"" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > next_replica_id:3
I161004 19:12:22.142867 60 storage/raft_transport.go:418  raft transport stream to node 1 established
I161004 19:12:22.146023 385 storage/replica.go:1712  [s1,r1:/M{in-ax}] proposing REMOVE_REPLICA {NodeID:1 StoreID:1 ReplicaID:1} for range 1: [{NodeID:2 StoreID:2 ReplicaID:2}]
I161004 19:12:22.147726 11 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I161004 19:12:22.147770 11 util/stop/stopper.go:425  quiescing; tasks left:
1      storage/replica_trigger.go:431
1      kv/txn_coord_sender.go:903
I161004 19:12:22.147827 11 util/stop/stopper.go:425  quiescing; tasks left:
1      storage/replica_trigger.go:431
I161004 19:12:22.147929 386 storage/replica_trigger.go:426  [s1,r1:/M{in-ax}] unable to gossip first range; hasLease=false, err=node unavailable; try another peer
I161004 19:12:22.168559 11 gossip/gossip.go:268  NodeID set to 1
I161004 19:12:22.169237 29 storage/replica_trigger.go:426  [s2,r1:/M{in-ax}] unable to gossip first range; hasLease=false, err=range 1: replica {2 2 2} not lease holder; <nil> is
I161004 19:12:22.169440 11 storage/store.go:908  [n1,s1,r1:/M{in-ax}]: added to replica GC queue
W161004 19:12:22.235137 361 storage/replica.go:3280  [s1,r1:/M{in-ax}] could not acquire lease for range gossip: range 1 was not found
W161004 19:12:22.235164 361 storage/store.go:1088  [s1] error gossiping first range descriptor: range 1 was not found
W161004 19:12:22.235317 362 storage/replica.go:3280  [s1] could not acquire lease for range gossip: range 1 was not found
W161004 19:12:22.235336 362 storage/store.go:1088  [s1] error gossiping system config: range 1 was not found
E161004 19:12:22.235840 467 storage/queue.go:560  [replicaGC] on [n1,s1,r1:/M{in-ax}]: no replica node addresses available via gossip
I161004 19:12:50.309770 144 gossip/gossip.go:1061  node has connected to cluster via gossip
I161004 19:12:58.253289 239 gossip/gossip.go:1061  node has connected to cluster via gossip
I161004 19:13:07.418250 517 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I161004 19:13:07.418496 517 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I161004 19:13:07.418697 517 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I161004 19:13:07.419004 136 http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:40629->127.0.0.1:39853: read: connection reset by peer
I161004 19:13:07.419017 45 http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:40067->127.0.0.1:41878: use of closed network connection
I161004 19:13:07.419057 517 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
I161004 19:13:07.419251 517 util/stop/stopper.go:353  stop has been called, stopping or quiescing all running tasks
--- FAIL: TestRemoveRangeWithoutGC (45.61s)
	testing.go:117: storage/client_raft_test.go:2360, condition failed to evaluate within 45s: expected replica to be missing
FAIL


ERROR: exit status 1

19786 runs completed, 1 failures, over 3m18s
FAIL
Makefile:146: recipe for target 'stress' failed
make: *** [stress] Error 1
```

--- a/pkg/storage/client_raft_test.go
+++ b/pkg/storage/client_raft_test.go
@@ -2441,7 +2441,7 @@ func TestRemoveRangeWithoutGC(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
 	sc := storage.TestStoreConfig(nil)
-	sc.TestingKnobs.DisableScanner = true
+	sc.TestingKnobs.DisableReplicaGCQueue = true
 	mtc := &multiTestContext{storeConfig: &sc}
 	mtc.Start(t, 2)
 	defer mtc.Stop()
@@ -2449,7 +2449,9 @@ func TestRemoveRangeWithoutGC(t *testing.T) {
 	mtc.replicateRange(rangeID, 1)
 	mtc.unreplicateRange(rangeID, 0)
 
-	// Wait for store 0 to process the removal.
+	// Wait for store 0 to process the removal. The in-memory replica
+	// object still exists but store 0 is no longer present in the
+	// configuration.
 	util.SucceedsSoon(t, func() error {
 		rep, err := mtc.stores[0].GetReplica(rangeID)
 		if err != nil {
@@ -2462,8 +2464,7 @@ func TestRemoveRangeWithoutGC(t *testing.T) {
 		return nil
 	})
 
-	// The replica's data is still on disk even though the Replica
-	// object is removed.
+	// The replica's data is still on disk.
 	var desc roachpb.RangeDescriptor
 	descKey := keys.RangeDescriptorKey(roachpb.RKeyMin)
 	if ok, err := engine.MVCCGetProto(context.Background(), mtc.stores[0].Engine(), descKey,
@@ -2473,16 +2474,29 @@ func TestRemoveRangeWithoutGC(t *testing.T) {
 		t.Fatal("expected range descriptor to be present")
 	}
 
-	// Stop and restart the store to reset the replica's raftGroup
-	// pointer to nil. As long as the store has not been restarted it
-	// can continue to use its last known replica ID.
+	// Stop and restart the store. The primary motiviation for this test
+	// is to ensure that the store does not panic on restart (as was
+	// previously the case).
 	mtc.stopStore(0)
 	mtc.restartStore(0)
 
+	// Initially, the in-memory Replica object is recreated from the
+	// on-disk state.
+	if _, err := mtc.stores[0].GetReplica(rangeID); err != nil {
+		t.Fatal(err)
+	}
+
+	// Re-enable the GC queue to allow the replica to be destroyed
+	// (after the simulated passage of time).
+	mtc.expireLeases()
+	mtc.manualClock.Increment(int64(storage.ReplicaGCQueueInactivityThreshold + 1))
+	mtc.stores[0].SetReplicaGCQueueActive(true)
+	mtc.stores[0].ForceReplicaGCScanAndProcess()
+
 	util.SucceedsSoon(t, func() error {
 		// The Replica object should be removed.
-		if _, err := mtc.stores[0].GetReplica(rangeID); err == nil {
-			return errors.Errorf("expected replica to be missing")
+		if _, err := mtc.stores[0].GetReplica(rangeID); !testutils.IsError(err, "range [0-9]+ was not found") {
+			return errors.Errorf("expected replica to be missing; got %v", err)
 		}
 
 		// And the data should no longer be on disk.
@@ -2490,7 +2504,7 @@ func TestRemoveRangeWithoutGC(t *testing.T) {
 			mtc.stores[0].Clock().Now(), true, nil, &desc); err != nil {
 			return err
 		} else if ok {
-			return errors.Errorf("expected range descriptor to be absent")
+			return errors.New("expected range descriptor to be absent")
 		}
 		return nil
 	})

--- a/pkg/storage/queue.go
+++ b/pkg/storage/queue.go
@@ -695,6 +695,16 @@ func (bq *baseQueue) remove(item *replicaItem) {
 // DrainQueue locks the queue and processes the remaining queued replicas. It
 // processes the replicas in the order they're queued in, one at a time.
 // Exposed for testing only.
+//
+// TODO(bdarnell): this method may race with the call to bq.pop() in
+// the main loop, in which case it does not guarantee that all
+// replicas have been processed by the time it returns. This is most
+// noticeable with ForceReplicaGCScanAndProcess, since the replica GC
+// queue has many event-driven triggers. This should synchronize
+// somehow with processLoop so we wait for anything being handled
+// there to finish too. When that's done, the SucceedsSoon at the end
+// of TestRemoveRangeWithoutGC (and perhaps others) can be replaced
+// with a one-time check.
 func (bq *baseQueue) DrainQueue(clock *hlc.Clock) {
 	ctx := bq.AnnotateCtx(context.TODO())
 	for repl := bq.pop(); repl != nil; repl = bq.pop() {

